overseer/src/lib.rs

```diff
@@ -0,0 +1,552 @@
+				write!(f, "OverseerMessage::SubsystemMessage{{ to: {:?}, msg: {:?} }}", to, msg)
+			}
+		    OverseerMessage::Spawn{ .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+    }
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+	/// The actual async task running.
+	f: SubsystemJob,
+}
+
+/// An `id` that is given to any `SubsystemInstance` for identification.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct SubsystemId(usize);
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M>>,
+}
+
+impl<M: Debug> SubsystemContext<M> {
+	/// Try to asyncronously receive a message.
+	///
+	/// This has to be used with caution, if you loop over this without
+	/// using `pending!()` macro you will end up with a busy loop!
+	pub async fn try_recv(&mut self) -> Result<Option<M>, ()> {
+		match poll!(self.rx.next()) {
+		    Poll::Ready(Some(msg)) => Ok(Some(msg)), 
+			Poll::Ready(None) => Err(()),
+		    Poll::Pending => Ok(None),
+		}
+	}
+
+	/// Receive a message.
+	pub async fn recv(&mut self) -> Result<M, SubsystemError> {
+		self.rx.next().await.ok_or(SubsystemError)
+	}
+
+	/// Send a message to whom it may concern.
+	///
+	/// The message will be broadcasted to all other `Subsystem`s that can
+	/// receive it.
+	pub async fn send_msg(&mut self, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
+			to: None,
+			msg,
+		}).await;
+	}
+
+	/// Spawn a child `Subsystem` on the executor and get it's `SubsystemId` upon success.
+	pub async fn spawn(&mut self, s: Box<dyn Subsystem<M>>) -> Result<SubsystemId, SubsystemError> {
+		let (tx, rx) = oneshot::channel();
+		let _ = self.tx.send(OverseerMessage::Spawn{
+			s,
+			res: tx,
+		}).await;
+
+		rx.await.unwrap_or_else(|_| Err(SubsystemError))
+	}
+
+	/// Send a direct message to some other `Subsystem` you know `SubsystemId` of.
+	pub async fn send_msg_to(&mut self, to: SubsystemId, msg: M) {
+		let _ = self.tx.send(OverseerMessage::SubsystemMessage{
+			to: Some(to),
+			msg,
+		}).await;
+	}
+
+	fn new(rx: mpsc::Receiver<M>, tx: mpsc::Sender<OverseerMessage<M>>) -> Self {
+		Self {
+			rx,
+			tx,
+		}
+	}
+}
+
+/// A trait that describes the `Subsystems` that can run on the `Overseer`.
+///
+/// It is generic over the message type circulating in the system.
+/// The idea that we want some type contaning persistent state that
+/// can start actually running jobs when asked to.
+pub trait Subsystem<M: Debug> {
+	/// Start this `Subsystem` and return `SubsystemJob`.
+	fn start(&mut self, ctx: SubsystemContext<M>) -> SubsystemJob;
```

Comment:
Naming the return type of this function a `SubsystemJob` is confusing, because it's really just a spawned subsystem. Maybe `SpawnedSubsystem` would be better.

Reply:
Additionally, I'd rather we get rid of this `SubsystemContext` type and just pass in the `Sender` and Receiver` separately. This way you'd be able to clone the sender etc.

Reply:
I don't believe we should need to clone the sender, as we generally want sub-task communication to go through the subsystem and then through the overseer.